File: elixir/day_04/day_04.ex
defmodule Day04 do
  defp read_input(input) do
    Path.join(["..", "..", "inputs", "day_04", input])
    |> File.read!()
    |> String.replace("\r\n", "\n")
    |> String.split("\n", trim: true)
  end

  defp convert_to_integer_list(string_list) do
    string_list
    |> Enum.map(&String.to_integer/1)
  end

  defp convert_to_sets(list_of_lists) do
    list_of_lists
    |> Enum.map(&MapSet.new/1)
  end

  defp intersect_sets(list_of_sets) do
    list_of_sets
    |> Enum.reduce(&MapSet.intersection(&1, &2))
  end

  defp process_line({line, index}) do
    line
    |> String.split(": ", trim: true)
    |> Enum.at(1)
    |> String.split(" | ", trim: true)
    |> Enum.map(&String.split(&1, " ", trim: true))
    |> Enum.map(&convert_to_integer_list/1)
    |> then(fn list -> {index, list} end)
  end

  defp win_copies(count_map, index_seq, total_cards) do
    case {index_seq, total_cards} do
      {[], ^total_cards} ->
        total_cards

      {[head | tail], ^total_cards} ->
        0..Map.get(count_map, head, 0)
        |> Enum.filter(&(&1 > 0))
        |> Enum.map(fn x -> head + x end)
        |> Enum.filter(fn x -> Map.has_key?(count_map, x) end)
        |> Enum.reverse()
        |> Enum.reduce(tail, fn x, acc ->
          [x | acc]
        end)
        |> then(fn acc ->
          {count_map, acc, total_cards + 1}
        end)
        |> then(fn {count_map, acc, total_cards} ->
          win_copies(count_map, acc, total_cards)
        end)
    end
  end

  def part_one do
    read_input("input.txt")
    |> Enum.with_index(1)
    |> Enum.map(&process_line/1)
    |> Enum.map(&elem(&1, 1))
    |> Enum.map(&convert_to_sets/1)
    |> Enum.map(&intersect_sets/1)
    |> Enum.map(&MapSet.size/1)
    |> Enum.filter(&(&1 > 0))
    |> Enum.map(fn size -> max(0, size - 1) end)
    |> Enum.map(fn size -> :math.pow(2, size) end)
    |> Enum.sum()
    |> trunc()
  end

  def part_two do
    read_input("input.txt")
    |> Enum.with_index(1)
    |> Enum.map(&process_line/1)
    |> Enum.map(fn {index, list} -> {index, convert_to_sets(list)} end)
    |> Enum.map(fn {index, sets} -> {index, intersect_sets(sets)} end)
    |> Enum.map(fn {index, intersections} -> {index, MapSet.size(intersections)} end)
    # Turn this into a map
    |> Enum.reduce(%{}, fn {index, size}, acc -> Map.put(acc, index, size) end)
    |> then(fn count_map ->
      win_copies(count_map, Map.keys(count_map) |> Enum.sort(), 0)
    end)
  end
end

Day04.part_one() |> IO.puts()
Day04.part_two() |> IO.puts()

File: golang/day_04/day_04.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strings"
)

type Card struct {
	winningNumbers map[string]bool
	scratchNumbers map[string]bool
}

func (c *Card) countWinningNumbers() int {
	sIntersection := map[string]bool{}

	for k := range c.winningNumbers {
		if _, ok := c.scratchNumbers[k]; ok {
			sIntersection[k] = true
		}
	}

	return len(sIntersection)
}

func readFile(filename string) []string {
	actualFilePath := filepath.Join("..", "..", "inputs", "day_04", filename)
	dat, err := os.ReadFile(actualFilePath)

	if err != nil {
		panic(err)
	}

	stringDat := string(dat)

	stringDat = strings.ReplaceAll(stringDat, "\r\n", "\n")

	splitFn := func(c rune) bool {
		return c == '\n'
	}

	return strings.FieldsFunc(stringDat, splitFn)
}

func parseLine(line string) Card {
	splitLine := strings.Split(line, ":")

	numbers := splitLine[1]

	splitNumbers := strings.Split(numbers, "|")

	winningNumbers := map[string]bool{}

	splitFn := func(c rune) bool {
		return c == ' '
	}

	for _, number := range strings.FieldsFunc(splitNumbers[0], splitFn) {
		winningNumber := strings.TrimSpace(number)
		winningNumbers[winningNumber] = true
	}

	scratchNumbers := map[string]bool{}

	for _, number := range strings.FieldsFunc(splitNumbers[1], splitFn) {
		scratchNumber := strings.TrimSpace(number)
		scratchNumbers[scratchNumber] = true
	}

	return Card{winningNumbers, scratchNumbers}
}

func part_one() {
	readData := readFile("input.txt")

	total := 0

	for _, line := range readData {
		card := parseLine(line)

		winningCount := card.countWinningNumbers()

		if winningCount < 1 {
			continue
		}

		total += int(math.Pow(2, float64(winningCount-1)))
	}

	fmt.Println("Part one:", total)
}

func pop(slice *[]int) (int, []int) {
	head := (*slice)[0]
	tail := (*slice)[1:]

	return head, tail
}

func winCopies(
	winCountMap *map[int]int,
	cards *[]int,
) int {
	total := 0

	for len(*cards) > 0 {
		head, tail := pop(cards)
		*cards = tail

		winCountForCard := (*winCountMap)[head]

		if winCountForCard == 0 {
			total += 1
			continue
		}

		wonCards := []int{}

		for i := 1; i <= winCountForCard; i++ {
			wonCard := head + i

			if _, ok := (*winCountMap)[wonCard]; ok {
				wonCards = append(wonCards, wonCard)
			} else {
				break
			}
		}

		*cards = append(wonCards, *cards...)
		total += 1
	}

	return total
}

func part_two() {
	readData := readFile("input.txt")

	winCountMap := map[int]int{}

	cards := []int{}

	for index, line := range readData {
		card := parseLine(line)

		winningCount := card.countWinningNumbers()

		winCountMap[index+1] = winningCount

		cards = append(cards, index+1)
	}

	totalCount := winCopies(&winCountMap, &cards)

	fmt.Println("Part two:", totalCount)

}

func main() {
	part_one()
	part_two()
}
